sdk/data/azcosmos/cosmos_client_retry_policy.go

```diff
@@ -0,0 +1,129 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package azcosmos
+
+import (
+	"fmt"
+	"net/http"
+	"time"
+
+	"github.com/Azure/azure-sdk-for-go/sdk/azcore/policy"
+)
+
+type clientRetryPolicy struct {
+	gem                    *globalEndpointManager
+	retryCount             int
+	sessionRetryCount      int
+	preferredLocationIndex int
+}
+
+const maxRetryCount = 10 // this number needs to be higher, keeping low for testing purposes right now
+const defaultBackoff = 1
+
+func (p *clientRetryPolicy) Do(req *policy.Request) (*http.Response, error) {
+	response, err := req.Next() // err can happen in weird scenarios (connectivity, etc) - need to test
+	o := pipelineRequestOptions{}
+	req.OperationValue(&o)
+	subStatus := response.Header.Get(cosmosHeaderSubstatus)
+	if p.shouldRetryStatus(response.StatusCode, subStatus) {
+		p.retryCount = 0
+		p.sessionRetryCount = 0
+		for {
+			resolvedEndpoint := p.gem.ResolveServiceEndpoint(p.retryCount, o.isWriteOperation)
```

Comment:
This should be before the `response, err := req.Next()`.

1. Resolve endpoint
2. Execute 
3. if shouldRetry
4. Mark endpoints unavailable
5. Resolve endpoint again (this time the endpoints are marked)
6. Retry request